Introduce Access Graph EventsStreamV2

jakule opened the pull request:
A new GRPC method EventsStreamV2 is introduced. The new method allows returning responses from the Access Graph. The new API is required by crown jewel implementation. The new API will be used to emit audit events from the access graph.

jakule committed:
Introduce Access Graph EventsStreamV2

A new GRPC method EventsStreamV2 is introduced. The new method allows returning responses from the Access Graph.
The new API is required by crown jewel implementation. The new API will be used to emit audit events from the access graph.

tigrato commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -6142,3 +6143,16 @@ message SessionRecordingConfigUpdate {
+  ];
+
+  // ChangeID is the id of the change.
+  string ChangeID = 2 [(gogoproto.jsontag) = "change_id"];
```
should we include more details in the audit log entry like affected resources,... so users can search them

jakule replied:
Added

tigrato commented on proto/accessgraph/v1alpha/access_graph_service.proto:
```diff
@@ -130,13 +136,56 @@ message EventsStreamRequest {
+message AuditEvent {
+  oneof event {
+    // access_path_changed is an event that should be logged when the access path changes.
+    events.AccessPathChanged access_path_changed = 1;
```
I would add more details here and maybe renaming it to be something different than audit event.
I think we should also handle the tag -> teleport conversion in teleport itself instead of having TAG sending the prepared audit event

jakule replied:
> I would add more details here 

What details?

> I think we should also handle the tag -> teleport conversion in teleport itself instead of having TAG sending the prepared audit event

I was also thinking about it, but then you end up with every audit event duplicated in two places.

tigrato replied:
What do you mean by every audit event duplicated?

jakule replied:
If we want to create an audit even we still need the audit event definition. If we want to have a separate definition for Teleport <-> TAG then we need a second definition that will have the same fields. 
I'm trying to say, that converting the audit events between Teleport - TAG sounds like a good idea, but this will require to create multiple proto messages with the same fields.

jakule replied:
>  return/add more details in one than what we have in the other.

@tigrato What is the advantage of returning more details from TAG and not recording them in the audit log?

tigrato replied:
For now we might not have, but imagine we are interested in collecting posthog metrics. It would become easier to do so if we had separate entities

jakule committed:
Add new fields to Event protobuf for better tracking

Added two new fields, AffectedResourceName and AffectedResourceSource, to the Event protobuf. These additions allow for a more detailed and efficient tracking of changes and their sources, improving overall monitoring and event handling.

tigrato commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -6142,3 +6143,22 @@ message SessionRecordingConfigUpdate {
+  string ChangeID = 2 [(gogoproto.jsontag) = "change_id"];
+
+  // AffectedResourceID is the name of the affected resource.
+  string AffectedResourceName = 3 [(gogoproto.jsontag) = "affected_resource_name"];
```
are we emitting an audit per resource?
Should we also include for `identities`?

jakule replied:
I'd skip it for now and if we decide to add it add it then. The identities can be tricky to get in some cases.

justinas replied:
> are we emitting an audit per resource?
👍 I'm also interested in this. I guess logic how to group the changes is non-obvious (by matcher?)

jakule replied:
I missed that.
I think we should emit one audit event per resource, that's why getting the user can be tricky as one resource modification can by caused by multiple events.

justinas reviewed (approved):
Looks good in general.

jakule committed:
Add AccessPathChanged event and update service file

A new proto file, events.proto, has been added to define the AccessPathChanged event, detailing changes in access paths. The access_graph_service.proto has been updated to use this new event instead of the legacy one. The relevant changes are also reflected in the generated Go file, access_graph_service.pb.go.

jakule committed:
Missing files
